File: algorithms/code/leetcode/lc017_letter_combinations_of_a_phone_number/LC017LetterCombinationsOfAPhoneNumberTests.java
package leetcode.lc017_letter_combinations_of_a_phone_number;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public final class LC017LetterCombinationsOfAPhoneNumberTests {
    @Test
    public void testNothing() throws Exception {
        LC017LetterCombinationsOfAPhoneNumber solution;
        solution = new LC017LetterCombinationsOfAPhoneNumber();
        List<String> result = solution.letterCombinations(null);
        assertEquals(0, result.size());
    }

    @Test
    public void testEmpty() throws Exception {
        LC017LetterCombinationsOfAPhoneNumber solution;
        solution = new LC017LetterCombinationsOfAPhoneNumber();
        List<String> result = solution.letterCombinations("");
        assertNotEquals(null, result);
        assertEquals(0, result.size());
    }

    @Test
    public void testExample() throws Exception {
        LC017LetterCombinationsOfAPhoneNumber solution;
        solution = new LC017LetterCombinationsOfAPhoneNumber();
        String[] expected = { "ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf" };
        Arrays.sort(expected);
        List<String> result = solution.letterCombinations("23");
        Collections.sort(result);
        assertNotEquals(null, result);
        assertEquals(expected.length, result.size());
        for (int i = 0; i < expected.length; i++) {
            assertEquals(expected[i], result.get(i));
        }
    }
}

File: algorithms/code/leetcode/lc017_letter_combinations_of_a_phone_number/LC017LetterCombinationsOfAPhoneNumber.Tests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Xunit;

namespace AlgorithmDesign.code.leetcode.lc017_letter_combinations_of_a_phone_number
{
    public class SolutionTests
    {
        [Fact]
        public void TestNothing()
        {
            IList<string> result = new Solution().LetterCombinations(null);
            Assert.Equal(0, result.Count);
        }

        [Fact]
        public void TestEmpty()
        {
            IList<string> result = new Solution().LetterCombinations("");
            Assert.Equal(0, result.Count);
        }

        [Fact]
        public void TestExample()
        {
            string[] expected = { "ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf" };
            Array.Sort(expected);
            IList<string> result = new Solution().LetterCombinations("23");
            List<string> sorted = result.ToList();
            sorted.Sort();
            Assert.Equal(expected.Length, sorted.Count);
            for (int i = 0; i < expected.Length; i++)
            {
                Assert.Equal(expected[i], sorted[i]);
            }
        }
    }
}
